File: baseproject-android/app/src/main/java/com/chang/template/activity/BaseActivity.java
package com.chang.template.activity;

import android.content.Intent;
import android.os.Build;
import android.os.Bundle;
import android.support.annotation.LayoutRes;
import android.support.annotation.Nullable;
import android.support.v4.app.NavUtils;
import android.support.v4.app.TaskStackBuilder;
import android.support.v4.content.ContextCompat;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;

import com.chang.template.R;
import com.chang.template.utils.Utility;

import butterknife.BindView;
import butterknife.ButterKnife;

/**
 * Created by Howard Chang on 2016/11/29
 */
public abstract class BaseActivity extends AppCompatActivity {

    @Nullable
    @BindView(R.id.toolbar)
    protected Toolbar toolbar;

    @LayoutRes
    protected abstract int getContentView();

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(getContentView());
        ButterKnife.bind(this);

        if (toolbar != null) {
            setSupportActionBar(toolbar);
        }

        displayHomeButton(true);

        initView();
    }

    private void initView() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT && Build.VERSION.SDK_INT < Build.VERSION_CODES.LOLLIPOP) {
            int color = ContextCompat.getColor(this, R.color.colorAccent);
            ViewGroup contentView = (ViewGroup) findViewById(android.R.id.content);
            View statusBarView = new View(this);
            ViewGroup.LayoutParams lp = new ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT,
                    Utility.getStatusBarHeight(this));
            statusBarView.setBackgroundColor(color);
            if (contentView != null) {
                contentView.addView(statusBarView, lp);
            }
        }
    }

    public void displayHomeButton(boolean isDisplay) {
        if (getSupportActionBar() != null) {
            getSupportActionBar().setDisplayHomeAsUpEnabled(isDisplay);
            getSupportActionBar().setHomeButtonEnabled(isDisplay);
        }
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                Intent upIntent = NavUtils.getParentActivityIntent(this);

                if ((upIntent != null && NavUtils.shouldUpRecreateTask(this, upIntent))
                        || isTaskRoot()) {
                    // This activity is NOT part of this app's task, so create a new task
                    // when navigating up, with a synthesized back stack.
                    TaskStackBuilder.create(this)
                            // Add all of this activity's parents to the back stack
                            .addNextIntentWithParentStack(upIntent)
                            // Navigate up to the closest parent
                            .startActivities();
                } else {
                    // This activity is part of this app's task, so simply
                    // navigate up to the logical parent activity.
                    onBackPressed();
                }

                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }

}

File: baseproject-android/app/src/main/java/com/chang/template/activity/BaseActivity.kt
package com.chang.template.activity

import android.os.Bundle
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import androidx.annotation.LayoutRes
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.NavUtils
import androidx.core.app.TaskStackBuilder
import androidx.core.content.ContextCompat
import com.chang.template.R
import com.chang.template.utils.Utility
import kotlinx.android.synthetic.main.view_toolbar.*

/**
 * Created by Howard Chang on 2016/11/29
 */
abstract class BaseActivity : AppCompatActivity() {

    @get:LayoutRes
    protected abstract val contentView: Int

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(contentView)

        setSupportActionBar(toolbar)

        displayHomeButton(true)

        initView()
    }

    private fun initView() {
        val color = ContextCompat.getColor(this, R.color.colorAccent)
        val contentView = findViewById<View>(android.R.id.content) as ViewGroup
        val statusBarView = View(this)
        val lp = ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT,
                Utility.getStatusBarHeight(this))
        statusBarView.setBackgroundColor(color)
        contentView.addView(statusBarView, lp)
    }

    fun displayHomeButton(isDisplay: Boolean) {
        supportActionBar?.setDisplayHomeAsUpEnabled(isDisplay)
        supportActionBar?.setHomeButtonEnabled(isDisplay)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> {
                NavUtils.getParentActivityIntent(this)?.apply {
                    if (NavUtils.shouldUpRecreateTask(this@BaseActivity, this) || isTaskRoot) {
                        // This activity is NOT part of this app's task, so create a new task
                        // when navigating up, with a synthesized back stack.
                        TaskStackBuilder.create(this@BaseActivity)
                                // Add all of this activity's parents to the back stack
                                .addNextIntentWithParentStack(this)
                                // Navigate up to the closest parent
                                .startActivities()
                    } else {
                        // This activity is part of this app's task, so simply
                        // navigate up to the logical parent activity.
                        onBackPressed()
                    }
                }

                return true
            }
            else -> return super.onOptionsItemSelected(item)
        }
    }

    companion object {
        val DATA = "data"
    }

}
